Test distributed collectives profiling with Gloo on GPU

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49072 Test distributed collectives profiling with Gloo on GPU**

As per the title, we should enable these tests for Gloo when run on GPU and the profiler is enabled with `use_cuda=True`. Enabling ProcessGroupNCCL profiling test to work with `use_cuda=True` is being tracked in https://github.com/pytorch/pytorch/issues/48987.

Also manually verified that it gives accurate results with GPU profiling, by timing the allreduce (which is blocking when we call wait() with gloo) with a vanilla `time.time()` and verified that the reported times are roughly similar.  

Also, we refactor some duplicated code in `test_reduce_multigpu` that makes it easier to read. 

Differential Revision: [D25388986](https://our.internmc.facebook.com/intern/diff/D25388986/)

rohan-varma committed:
Test distributed collectives profiling with Gloo on GPU

As per the title, we should enable these tests for Gloo when run on GPU and the profiler is enabled with `use_cuda=True`. Enabling ProcessGroupNCCL profiling test to work with `use_cuda=True` is being tracked in https://github.com/pytorch/pytorch/issues/48987.

Differential Revision: [D25388986](https://our.internmc.facebook.com/intern/diff/D25388986/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Test distributed collectives profiling with Gloo on GPU"


As per the title, we should enable these tests for Gloo when run on GPU and the profiler is enabled with `use_cuda=True`. Enabling ProcessGroupNCCL profiling test to work with `use_cuda=True` is being tracked in https://github.com/pytorch/pytorch/issues/48987.

Differential Revision: [D25388986](https://our.internmc.facebook.com/intern/diff/D25388986/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Test distributed collectives profiling with Gloo on GPU"


As per the title, we should enable these tests for Gloo when run on GPU and the profiler is enabled with `use_cuda=True`. Enabling ProcessGroupNCCL profiling test to work with `use_cuda=True` is being tracked in https://github.com/pytorch/pytorch/issues/48987.

Differential Revision: [D25388986](https://our.internmc.facebook.com/intern/diff/D25388986/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Test distributed collectives profiling with Gloo on GPU"


As per the title, we should enable these tests for Gloo when run on GPU and the profiler is enabled with `use_cuda=True`. Enabling ProcessGroupNCCL profiling test to work with `use_cuda=True` is being tracked in https://github.com/pytorch/pytorch/issues/48987.

Differential Revision: [D25388986](https://our.internmc.facebook.com/intern/diff/D25388986/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Test distributed collectives profiling with Gloo on GPU"


As per the title, we should enable these tests for Gloo when run on GPU and the profiler is enabled with `use_cuda=True`. Enabling ProcessGroupNCCL profiling test to work with `use_cuda=True` is being tracked in https://github.com/pytorch/pytorch/issues/48987.

Differential Revision: [D25388986](https://our.internmc.facebook.com/intern/diff/D25388986/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Test distributed collectives profiling with Gloo on GPU"


As per the title, we should enable these tests for Gloo when run on GPU and the profiler is enabled with `use_cuda=True`. Enabling ProcessGroupNCCL profiling test to work with `use_cuda=True` is being tracked in https://github.com/pytorch/pytorch/issues/48987.

Differential Revision: [D25388986](https://our.internmc.facebook.com/intern/diff/D25388986/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Test distributed collectives profiling with Gloo on GPU"


As per the title, we should enable these tests for Gloo when run on GPU and the profiler is enabled with `use_cuda=True`. Enabling ProcessGroupNCCL profiling test to work with `use_cuda=True` is being tracked in https://github.com/pytorch/pytorch/issues/48987.

Differential Revision: [D25388986](https://our.internmc.facebook.com/intern/diff/D25388986/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Test distributed collectives profiling with Gloo on GPU"


As per the title, we should enable these tests for Gloo when run on GPU and the profiler is enabled with `use_cuda=True`. Enabling ProcessGroupNCCL profiling test to work with `use_cuda=True` is being tracked in https://github.com/pytorch/pytorch/issues/48987.

Differential Revision: [D25388986](https://our.internmc.facebook.com/intern/diff/D25388986/)

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -1332,6 +1347,24 @@ def _test_all_reduce_helper(
+                        op,
+                        group_id,
+                        async_op=async_op,
+                        profile_cuda=True,
```
Would I be correct if I assume this will only make a difference when `cuda == True`? Is there any existing tests exploring that code path?

rohan-varma replied:
`cuda` controls whether GPU tensors are used instead of CPU tensors, but `profile_cuda` controls whether profiling is enabled with CUDA or not, so they are sort of independent.

mrshenli commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -1289,12 +1294,22 @@ def test_all_reduce_result_cuda(self):
                 op_calls.append(secondary_op_call)
 
-            with torch.autograd.profiler.profile() as prof:
+            with torch.autograd.profiler.profile(use_cuda=profile_cuda) as prof:
```
qq, does `use_cuda` change any info on the event? Is there a way to verify that `use_cuda` does make a difference here?

rohan-varma replied:
It doesn't change any info on the `gloo::allreduce` or other collectives since they are profiled with `RecordFunction`, although the aten ops such as `copy` are appropriately recorded with cuda time. I think asserting on that is better covered by profiler specific tests though, here we just want to ensure that enabling cuda profiling does not cause any issues.

mrshenli reviewed (approved):
code change looks good to me.

rohan-varma committed:
Update on "Test distributed collectives profiling with Gloo on GPU"


As per the title, we should enable these tests for Gloo when run on GPU and the profiler is enabled with `use_cuda=True`. Enabling ProcessGroupNCCL profiling test to work with `use_cuda=True` is being tracked in https://github.com/pytorch/pytorch/issues/48987.

Also manually verified that it gives accurate results with GPU profiling, by timing the allreduce (which is blocking when we call wait() with gloo) with a vanilla `time.time()` and verified that the reported times are roughly similar.  

Also, we refactor some duplicated code in `test_reduce_multigpu` that makes it easier to read. 

Differential Revision: [D25388986](https://our.internmc.facebook.com/intern/diff/D25388986/)

[ghstack-poisoned]
